// prevent-parent-scroll/dist/index.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var PreventParentScroll = function () {
    function PreventParentScroll(elem) {
        _classCallCheck(this, PreventParentScroll);

        this.elem = elem;
    }

    _createClass(PreventParentScroll, [{
        key: 'start',
        value: function start() {
            this.handler = this.onMouseWheel.bind(this);
            this.elem.addEventListener('mousewheel', this.handler, false);
        }
    }, {
        key: 'stop',
        value: function stop() {
            this.elem.removeEventListener('mousewheel', this.handler);
        }
    }, {
        key: 'onMouseWheel',
        value: function onMouseWheel(event) {
            if (event.preventParentScroll) {
                return;
            }
            var y = this.getDelta(event);
            if (y) {
                var outerHeight = this.elem.getBoundingClientRect().height;
                var computedStyle = window.getComputedStyle(this.elem);
                var borderTopWidth = parseFloat(computedStyle.borderTopWidth || '') || 0;
                var borderBottomWidth = parseFloat(computedStyle.borderBottomWidth || '') || 0;
                var elemScrollableHeight = outerHeight - borderTopWidth - borderBottomWidth;
                this.maybePreventMouseWheel(y, elemScrollableHeight, this.elem.scrollTop, this.elem.scrollHeight, event);
            }
            event.preventParentScroll = true;
        }
    }, {
        key: 'getDelta',
        value: function getDelta(event) {
            var xAxis = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

            if (event.wheelDelta) {
                var delta = event.wheelDeltaY;
                if (xAxis) {
                    delta = event.wheelDeltaX;
                }
                return delta;
            } else if (event.detail) {
                if (event.axis && (event.axis === 1 && xAxis || event.axis === 2 && !xAxis)) {
                    return -1 * event.detail * 12;
                }
            } else if (event.deltaMode) {
                if (xAxis) {
                    return event.deltaX;
                } else {
                    return event.deltaY;
                }
            }
            return 0;
        }
    }, {
        key: 'maybePreventMouseWheel',
        value: function maybePreventMouseWheel(delta, elemLength, elemScroll, scrollLength, event) {
            if (delta < 0 && this.scrollIsMaxed(elemLength, elemScroll, scrollLength)) {
                event.preventDefault();
            } else if (delta > 0 && this.scrollIsMinned(elemScroll)) {
                event.preventDefault();
            }
        }
    }, {
        key: 'scrollIsMaxed',
        value: function scrollIsMaxed(elemLength, elemScroll, scrollLength) {
            return elemLength + elemScroll === scrollLength;
        }
    }, {
        key: 'scrollIsMinned',
        value: function scrollIsMinned(elemScroll) {
            return elemScroll === 0;
        }
    }]);

    return PreventParentScroll;
}();

Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PreventParentScroll;
